### PR TITLE
Makefile setup, test stubs

### PR DESCRIPTION
**Pull request by michaelkleinhenz**
This adds a Makefile setup and stubs for unit and e2e tests. Execute `make` to get a list of possible targets. This was essentially taken from `toolchain-operator` with added test setup.

### CODE REVIEW

File: Dockerfile.builder [hunk 1]
```diff
@@ -0,0 +1,37 @@
+FROM centos:7
```
Comment (kwk) [inline Dockerfile.builder]: Instead of having two Dockerfiles, we should start using multi-staged Dockerfiles.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+CUR_DIR=$(shell pwd)
+TMP_PATH=$(CUR_DIR)/tmp
+INSTALL_PREFIX=$(CUR_DIR)/bin
+VENDOR_DIR=vendor
+INCLUDE_DIR=make
+SOURCE_DIR ?= .
+SOURCES := $(shell find $(SOURCE_DIR) -path $(SOURCE_DIR)/vendor -prune -o -name '*.go' -print)
+
+ifeq ($(OS),Windows_NT)
+	include ./$(INCLUDE_DIR)/Makefile.win
```
Comment (kwk) [inline Makefile]: Please, let's ignore Windows, okay? Nobody can test this code at the moment.

File: cmd/manager/main.go [hunk 3]
```diff
@@ -53,15 +53,25 @@ func main() {
 	}
 
 	// Become the leader before proceeding
-	leader.Become(context.TODO(), "devconsole-lock")
+	err = leader.Become(context.TODO(), "devconsole-lock")
+	if err != nil {
+		log.Error(err, "")
```
Comment (kwk) [inline cmd/manager/main.go]: Don't we have a `log.Panic` maybe?

**Comment by michaelkleinhenz** [comment]
@kwk no, all of the code is used/tested/inspected with the exception of the coverage and the windows stuff. There is not much overhead and I have gone over the make process line-by-line. @dipak-pawar did a great job there.

File: cmd/manager/main.go (hunk 3 above)
Comment (michaelkleinhenz) [reply cmd/manager/main.go]: I only updated this to make it compliant to the linter.

File: Makefile (hunk 2 above)
Comment (michaelkleinhenz) [reply Makefile]: @alexeykazakov @dipak-pawar I sort of agree with @kwk. Was there a specific reason to have Windows support in the toolchain operator? We should be consistent here.

Comment (dipak-pawar) [reply Makefile]: there is  no specific reason, also it was not tested by anyone, we can remove this. it was copy pasted when I wrote makefile, but we can get rid out of that

Comment (alexeykazakov) [reply Makefile]: I'm fine with removing Windows for now.

**Comment by sbose78** [comment]
Opened a PR to make the code build https://github.com/redhat-developer/devopsconsole-operator/pull/4

**Comment by kwk** [comment]
> @kwk no, all of the code is used/tested/inspected with the exception of the coverage and the windows stuff. There is not much overhead and I have gone over the make process line-by-line. @dipak-pawar did a great job there.

That is my point, the coverage stuff takes up 90% of the code and can be simplified/removed like crazy.

File: Makefile (hunk 2 above)
Comment (michaelkleinhenz) [reply Makefile]: Done. 98d2b3f

File: Gopkg.lock [hunk 4]
```diff
@@ -2,381 +2,387 @@
...
-  digest = "1:8098cd40cd09879efbf12e33bcd51ead4a66006ac802cd563a66c4f3373b9727"
   name = "github.com/PuerkitoBio/urlesc"
   packages = ["."]
-  pruneopts = "NT"
   revision = "de5bf2ad457846296e2031421a34e2568e304e35"
 
 [[projects]]
-  digest = "1:4b8b5811da6970495e04d1f4e98bb89518cc3cfc3b3f456bdb876ed7b6c74049"
+  branch = "master"
+  name = "github.com/alexeykazakov/devconsole"
```
[the remark this answers is not in the document]
Comment (michaelkleinhenz) [reply Gopkg.lock]: Yeah, rebased. There are also some base image references in the dockerfiles that need updating once we know the repository address.

**Comment by michaelkleinhenz** [comment]
@kwk Coverage is now also tested and working. Removed/updated a few things, also removed windows support. 98d2b3f.

File: deploy/test/global-manifests.yaml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+---
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: devconsole
+rules:
+- apiGroups:
+  - rbac.authorization.k8s.io
+  resources:
+  - clusterrolebindings
```
Comment (dipak-pawar) [inline deploy/test/global-manifests.yaml]: I think we don't need our test `sa` needs to have permission for `clusterrolebindings`

File: deploy/test/global-manifests.yaml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  verbs:
+  - '*'
+- apiGroups:
+  - oauth.openshift.io
+  resources:
+  - oauthclients
```
Comment (dipak-pawar) [inline deploy/test/global-manifests.yaml]: we don't need our test `sa` needs to have permission for `oauthclients`. this needs for toolchain, maybe copy pasted?

File: deploy/test/global-manifests.yaml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  verbs:
+  - '*'
+- apiGroups:
+  - ""
+  - project.openshift.io
+  resources:
+  - projectrequests
```
Comment (dipak-pawar) [inline deploy/test/global-manifests.yaml]: same as above

File: Dockerfile.deploy [hunk 8]
```diff
@@ -0,0 +1,18 @@
+FROM centos:7
+LABEL maintainer "Devtools <devtools@redhat.com>"
+LABEL author "Devtools <devtools@redhat.com>"
+ENV LANG=en_US.utf8
+ENV F8_INSTALL_PREFIX=/usr/local/devconsole-operator
```
Comment (dipak-pawar) [inline Dockerfile.deploy]: maybe let's get rid out of `F8` prefix

File: Dockerfile.deploy [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+# Create a non-root user and a group with the same name: "devconsole-operator"
+ENV F8_USER_NAME=devconsole-operator
```
Comment (dipak-pawar) [inline Dockerfile.deploy]: maybe let's get rid out of `F8` prefix

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,272 @@
+PROJECT_NAME=devopsconsole-operator
+PACKAGE_NAME:=github.com/fabric8-services/$(PROJECT_NAME)
```
Comment (dipak-pawar) [inline Makefile]: this should be `github.com/redhat-developer`

File: Makefile [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+# -------------------------------------------------------------------
+# Docker build
+# -------------------------------------------------------------------
+BUILD_DIR = bin
+REGISTRY_URI = quay.io
+REGISTRY_NS = ${PROJECT_NAME}
+REGISTRY_IMAGE = ${PROJECT_NAME}
+
+ifeq ($(TARGET),rhel)
+	REGISTRY_URL := ${REGISTRY_URI}/openshiftio/rhel-${REGISTRY_NS}-${REGISTRY_IMAGE}
```
Comment (dipak-pawar) [inline Makefile]: can we change docker org as we are starting fresh?

File: make/Makefile.dev [hunk 12]
```diff
@@ -0,0 +1,89 @@
+DOCKER_REPO?=quay.io/openshiftio
```
Comment (dipak-pawar) [inline make/Makefile.dev]: can we change this docker org?

File: make/Makefile.dev [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+	@oc delete -f $(DEPLOY_DIR)/cluster_role.yaml || true
+	@echo "Deleting Service Account"
+	@oc delete -f $(DEPLOY_DIR)/service_account.yaml || true
+	@echo "Deleting Custom Resources..."
+	@oc delete -f $(DEPLOY_DIR)/crds/devconsole_v1alpha1_envdeployment_crd.yaml|| true
+	@echo "Deleting Namespace"
+	@oc delete -f $(DEPLOY_DIR)/namespace.yaml || true
+	@echo "Deleting Custom Resource Definitions..."
+	@oc delete -f $(DEPLOY_DIR)/crds/devconsole_v1alpha1_envdeployment_crd.yaml || true
+	@echo "Deleting OAuthClient 'codeready-devconsole'"
```
Comment (dipak-pawar) [inline make/Makefile.dev]: we don't need `OAuthClient` here

File: make/test.mk [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+	go test ./test/e2e/... -root=$(PWD) -kubeconfig=$(HOME)/.kube/config -globalMan deploy/test/global-manifests.yaml -namespacedMan deploy/test/namespace-manifests.yaml -v -parallel=1 -singleNamespace
+
+.PHONY: e2e-setup
+e2e-setup:  e2e-cleanup
+	oc new-project devconsole-e2e-test || true
+
+.PHONY: e2e-cleanup
+e2e-cleanup:
+	oc login -u system:admin
+	oc delete oauthclient codeready-devconsole || true
```
Comment (dipak-pawar) [inline make/test.mk]: this can be removed

File: make/test.mk [hunk 15]
```diff
@@ -0,0 +1,350 @@
...
+	oc delete clusterrolebinding devconsole-enabler || true
```
Comment (dipak-pawar) [inline make/test.mk]: this can be removed

File: make/test.mk [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+	oc delete clusterrole devconsole-enabler || true
```
Comment (dipak-pawar) [inline make/test.mk]: this can be removed

File: test/README.md [hunk 17]
```diff
@@ -0,0 +1,27 @@
+## E2E Tests
+
+E2E tests are verifying successful deployment of the DevOps Console Operator and creation of self-provisioning service account.
```
Comment (dipak-pawar) [inline test/README.md]: creation of self-provisioning account is functionality for toolchain operator. we should remove this.

Comment (michaelkleinhenz) [reply test/README.md]: Fixed befe3ef.

File: make/test.mk (hunk 16 above)
Comment (michaelkleinhenz) [reply make/test.mk]: Fixed befe3ef.

File: make/test.mk (hunk 15 above)
Comment (michaelkleinhenz) [reply make/test.mk]: Fixed befe3ef.

File: make/test.mk (hunk 14 above)
Comment (michaelkleinhenz) [reply make/test.mk]: Fixed befe3ef.

File: make/Makefile.dev (hunk 13 above)
Comment (michaelkleinhenz) [reply make/Makefile.dev]: Fixed befe3ef.

File: make/Makefile.dev (hunk 12 above)
Comment (michaelkleinhenz) [reply make/Makefile.dev]: Yes, let's do this when we know what it is..

File: Makefile (hunk 11 above)
Comment (michaelkleinhenz) [reply Makefile]: Same as below. Let's do this when we know how it will look like.

File: Makefile (hunk 10 above)
Comment (michaelkleinhenz) [reply Makefile]: Fixed befe3ef.

File: Dockerfile.deploy (hunk 9 above)
Comment (michaelkleinhenz) [reply Dockerfile.deploy]: Fixed befe3ef.

File: Dockerfile.deploy (hunk 8 above)
Comment (michaelkleinhenz) [reply Dockerfile.deploy]: Fixed befe3ef.

File: deploy/test/global-manifests.yaml (hunk 7 above)
Comment (michaelkleinhenz) [reply deploy/test/global-manifests.yaml]: Fixed befe3ef.

File: deploy/test/global-manifests.yaml (hunk 6 above)
Comment (michaelkleinhenz) [reply deploy/test/global-manifests.yaml]: Fixed befe3ef.

File: deploy/test/global-manifests.yaml (hunk 5 above)
Comment (michaelkleinhenz) [reply deploy/test/global-manifests.yaml]: Fixed befe3ef.

File: Dockerfile.deploy (hunk 9 above)
Comment (kwk) [reply Dockerfile.deploy]: @michaelkleinhenz no it wasn't fixed in befe3ef. The code is still in there. We have `F8_` in various other locations and files as well if you take a look at the PR now. I suggest to keep the `F8_` and not replace it now. The reason is that environment variables marked with `F8_` are those that you should care about setting from the outside when running a Makefile for example: `F8_POSTGRES_HOST=foo make test-integration`. Can you revert the parts of befe3ef that modify the `F8_` bits please and keep the rest until we have a proper replacement for `F8_`?

**Comment by alexeykazakov** [comment]
@michaelkleinhenz any updates on that PR?

**Pull request merged**